src/ai.cpp

```diff
@@ -0,0 +1,321 @@
+			}
+		}
+	}
+}
+
+direction AI::make_villain_decide(direction active_lane, std::map<direction, Lane*> lanes) {
+	PlayerOutcome bestPlayerOutcome = { 
+		0,
+		direction::NORTH, // This will never be the outcome because the player will always move at least one car
+	};
+
+	get_best_turning_direction(active_lane, lanes, 0, 0, false, &bestPlayerOutcome);
+
+	return bestPlayerOutcome.most_villainous_direction;
+}
+
+void AI::get_best_turning_direction(
+	direction active_lane, 
+	std::map<direction, Lane*> current_lanes,
+	int current_ply, 
+	int cars_turned, 
+	bool villain_at_front, 
+	PlayerOutcome *bestPlayerOutcome) 
+{
+	if (villain_at_front) {
+		// The villain is now at the front of its lane. Time to decide the most dastardly villainous move for this scenario!
+
+		// If current_cars_turned_per_move is not better than the best outcome found so far, then the player won't do this
+		// so there's no reason for us to do any calculations here
+		float current_cars_turned_per_move = (float)cars_turned / (float)current_ply;
+		if (current_cars_turned_per_move > bestPlayerOutcome->cars_turned_per_move) {
+			bestPlayerOutcome->cars_turned_per_move = current_cars_turned_per_move;
+			bestPlayerOutcome->most_villainous_direction = get_most_villainous_direction(active_lane, current_lanes);
+		}
+
+		return;
+	}
+
+	if (current_ply == MaxPly) {
+		// We've reached our ply depth without moving our villain to the front of the line. If this is the best move
+		// for the player then it doesn't matter what we set our villain to... If we set our MaxPly to infinity then
+		// this will never occur and eventually the villain will reach the front of the line, but that would be
+		// computationally exhausting and, in fact, wouldn't terminate.
+
+		// If current_cars_turned_per_move is not better than the best outcome found so far, then the player won't do this
+		// so there's no reason for us to do any calculations here
+		float current_cars_turned_per_move = (float)cars_turned / (float)current_ply;
+		if (current_cars_turned_per_move > bestPlayerOutcome->cars_turned_per_move) {
+			bestPlayerOutcome->cars_turned_per_move = current_cars_turned_per_move;
+			bestPlayerOutcome->most_villainous_direction = direction(rand() % 4);
+		}
+
+		return;
+	}
+
+	// Don't want to modify the real game lanes
+	std::map<direction, Lane*> new_lanes = deep_copy_lanes(current_lanes);
+	std::vector<std::set<direction>> possible_moves = generate_moves(new_lanes);
+	
+	for (std::vector<std::set<direction>>::iterator it = possible_moves.begin(); it != possible_moves.end(); it++)
+	{
+		apply_move_to_lanes(*it, new_lanes);
+		bool is_moving_active_lane = it->find(active_lane) != it->end();
+		get_best_turning_direction(active_lane, new_lanes, current_ply++, cars_turned + it->size(), is_moving_active_lane, bestPlayerOutcome);
+		new_lanes = deep_copy_lanes(current_lanes);
+	}
+}
+
+std::map<direction, Lane*> AI::deep_copy_lanes(std::map<direction, Lane*> lanes) {
```

Comment:
I'd love for someone to verify that this is a good implementation of deepcopy in c++.

Reply:
(this seems to work)